### ISSUE TITLE
Support $redirect-urls in adblock

### ISSUE THREAD

**Issue by ShivanKaul** [brave/brave-browser#18542]
Related to https://github.com/brave/brave-browser/issues/18186. We've added support for `redirect-url` for adblock-rs which will allow us to specify loading a replacement resource over the network.
We will restrict this to known safe filter lists (disabled by default).

The intention is to allow resource replacements over the network if there's an adblock filter match. In other words, the redirect-filter option in a filter rule instructs the browser to load a replacement resource for the matching resource via a URL instead of locally. Example:

```
google-analytics.js$redirect-url=important,https://pcdn.brave.software/sugarcoat/sugarcoat-analytics-a14946c73def6bdaa93155f58435a82e.min.js
```
would load the script at [https://pcdn.brave.software/sugarcoat/sugarcoat-analytics-a14946c73def6bdaa93155f58435a82e.min.js](https://pcdn.brave.software/sugarcoat/sugarcoat-analytics-a14946c73def6bdaa93155f58435a82e.min.js) instead of any script named `google-analytics.js`. 

Notice the `pcdn.brave.software` domain - the intended usecase is to only load sugarcoated scripts hosted on [Private CDN](https://github.com/brave/brave-browser/wiki/Private-CDN). The only filter lists we will enable `redirect-url` filter option to be parsed on are Brave-controlled ones, and specifically only Sugarcoat for now.

Sugarcoat is a research project to automatically stub out privacy-invasive API accesses by scripts. For more info: https://github.com/brave-experiments/sugarcoat-pipeline/. Essentially, the sugarcoat pipeline takes in a webpage and gives you sanitized scripts that _would have been blocked by a filter list rule but are not because of webcompat reasons i.e. have a corresponding exception rule_. The generated sugarcoat rule has an `important` filter option so as to ensure the resource does get blocked despite the exception rule. We previously tried loading the sugarcoated replacement resources locally but that requires us to ship them in the browser which resulted in some memory-bound machines getting borked (this also has to do with some tech debt in how we load filter lists but in any case, for Sugarcoat to be useful we would need to load many, many replacement scripts). 

This feature should have a feature flag so we can turn it off via Griffin, though given that the sugarcoat rules are shipped in the adblock crx component we can just ship a component update which would not require a restart. 

Related PRs:
- Add sugarcoat rules (actually an update because of aforementioned-deploy that we rolled back in https://github.com/brave/adblock-resources/pull/32): https://github.com/brave/adblock-lists/pull/660
- Add sugarcoat list to default.json: https://github.com/brave/adblock-resources/pull/37

Security/privacy review: https://github.com/brave/security/issues/615

**Comment by stephendonner** [comment]
Can we get a (more-complete) test plan for this feature, @ShivanKaul?  Thanks!

**Comment by ShivanKaul** [comment]
@stephendonner updated the PR test plan and DM'd the DAT file for testing convenience.

**Comment by stephendonner** [comment]
> @stephendonner updated the PR test plan and DM'd the DAT file for testing convenience.

Thanks a bunch!

**Comment by stephendonner** [comment]
### Verification `IN-PROGRESS` using



Brave | 1.33.75 Chromium: 96.0.4664.35 (Official Build) beta (x86_64)
-- | --
Revision | 8df3babf8da8faa8aacb78f4848ef66a93b4944b-refs/branch-heads/4664@{#680}
OS | macOS Version 11.6.1 (Build 20G224)



### Steps:


1. new profile
2. renamed `rs-ABPFilterParserData.dat` in `/Users/stephendonner/Library/Application Support/BraveSoftware/Brave-Browser-Nightly/cffkpbalmllkdoenhmdmpbkajipdjfam/1.0.1121` to `rs-ABPFilterParserData.dat.ORIG`
3. dropped in the replacement `rs-ABPFilterParserData.dat` from @shivankaul
4. launch Brave
5. set `brave://flags/#brave-adblock-redirect-url` to `Enabled` 
6. open the Developer Tools
7. open the `Network` panel
8. load `vimeo.com`
9. load `reuters.com` 
10. load `bloomberg.com` 
11. load `researchgate.net` 
12. load `amazon.jobs`
13. load `fedex.com`
14. load `bmw.com.au` 
15. load `virginmobile.ca`
16. load `einthusan.tv`
17. load `downdetector.com`
18. for each of the above, verify the original request for the asset for that domain from `https://github.com/brave/adblock-lists/blob/master/brave-lists/brave-sugarcoat.txt` are replaced with their corresponding `https://pcdn.brave.com/sugarcoat/async-sugarcoat-[unique-string].js` file.

vimeo | reuters.com | bloomberg.com | researchgate.net| amazon.jobs | fedex.com | bmw.com.au | virginmobile.ca | einthusan.tv | downdetector.com
------|------|-----|-----|------|------|------|------|------|-----
<img width="1541" alt="Screen Shot 2021-11-10 at 10 15 04 AM" src="https://user-images.githubusercontent.com/387249/141311190-c64bd8ff-13c5-4b89-b4ae-e91d10f14a35.png"> | <img width="1541" alt="Screen Shot 2021-11-10 at 10 14 19 AM" src="https://user-images.githubusercontent.com/387249/141311320-d57f8bd2-ab99-42b7-b756-0440718b95cb.png"> | bloomberg | 
<img width="1541" alt="Screen Shot 2021-11-10 at 10 00 21 AM" src="https://user-images.githubusercontent.com/387249/141312914-8dfe154c-3cfe-46c1-b19f-de8abab23aaa.png"> | <img width="1541" alt="Screen Shot 2021-11-10 at 9 44 18 AM" src="https://user-images.githubusercontent.com/387249/141311878-fad5bea2-a0aa-4582-b642-ebc4842c9b35.png"> | <img width="1541" alt="Screen Shot 2021-11-10 at 10 26 53 am" src="https://user-images.githubusercontent.com/387249/141311751-62e52ebe-d389-4b25-a6f5-360b92ca8351.png"> | bmw.com.au | virginmobile.ca | <img width="1541" alt="Screen Shot 2021-11-10 at 9 42 06 AM" src="https://user-images.githubusercontent.com/387249/141312539-922804a0-50b0-4336-949d-135d602ea17f.png"> | <img width="1541" alt="Screen Shot 2021-11-10 at 10 18 10 AM" src="https://user-images.githubusercontent.com/387249/141311688-88b12fad-f6a7-4d0b-b60a-863f189efb34.png">

**Comment by stephendonner** [comment]
@ShivanKaul when you get a chance, can you please check on the `bloomberg.com`, `bmw.com.au`, and `virginmobile.ca` cases, as I think they've changed; thanks!

**Comment by stephendonner** [comment]
@ShivanKaul this has `OS/Android`, but can it really be tested on Android?  /cc @srirambv

**Comment by stephendonner** [comment]
> @ShivanKaul this has `OS/Android`, but can it really be tested on Android? /cc @srirambv

in talking with @samartnik and others, this isn't possible on opt Android builds, so removing `OS/Android` - if I'm wrong, please add back, of course

**Comment by ShivanKaul** [comment]
@samartnik I would assume Android also needs a DAT file similar to desktop for its adblocking? There's no way to load a custom one?

**Comment by samartnik** [comment]
@ShivanKaul not on non-rooted devices, these files are protected. We may add an option to `QA Preferences` to export/import adblocking dat file, like we do with rewards database, but that's a separate task.

**Comment by ShivanKaul** [comment]
@stephendonner do you have a rooted Android device, by any chance?

**Comment by stephendonner** [comment]
> @stephendonner do you have a rooted Android device, by any chance?

OEM-unlocked, but not rooted, no.  Don't believe anybody on the team runs with them, typically.
